Gracie/dashboard setup

artsygracie17 opened the pull request:
I renamed the branch to `gracie/dashboard-setup` because I realized I had a lot of code to review on this branch already concerning the API communication, transition between screens and other updates leading to and from the dashboard without actually having designed the dashboard yet. I will work on the actual dashboard screen on a separate branch.

Commit:
API client framework working and getCurrentUser working. Integrate Google sign in with added View Model; take things out from App Delegate

API client framework working and getCurrentUser working. Integrate Google sign in with added View Model; take things out from App Delegate

Add project pbx

Request, LucienRequest, APIClient, LucienAPIClient framework

Add getCurrentUser endpoint; working

Commit:
Create Dashboard and Dashboard navigation controller. Link profile to navigation smiley icon in dashboard. Decode token in separate AuthenticationResponse model.

Add setNavBarBackground and setNavBarTitle to LucienTheme to be accessed and customized throughout application whenever nav bar needs to be created

Restructure UINavigationController. Navigation flow from Dashboard to Profile working

Authorization token now set properly. Authenticate user everytime at sign in even if already signed in

Commit changes; in middle of getting getCurrentUser endpoint working

getCurrentUser working. Add AuthenticationResponse model for authenticate user and use User model for getCurrentUser only.

Add sample image

Commit:
Begin overall view layout of Dashboard. Add three views for lending borrowing and my collection (now empty). Horizontal scroll view working but nothing in it

Commit changes to checkout. Start UICollectionView

Change collection count decoder coding key to int so that Dashboard gets decoded properly

Grab Dashboard data successfully! So far only decoding myCollectionCount. Add logic to sign in RootViewController to bring user to either start a new collection or to the dashboard depending on myCollectionCount

Commit:
Update StartMyCollection design, components and constraints to match new Zeplin design; include linking to Dashboard

Update StartMyCollection xib design to Zeplin state

Add UILabel extension for letter spacing; StartMyCollection welcome screen UI elements updated

Edit constraints in home screen for aspect ratios conformity on iPhone touch

Tweak StartMyCollection screen for device demo

Commit:
Substitute Dashboard endpoint to has_collection so that the check of whether user has comics in collection or not runs faster

hasCollection with String intrepretation working (was in bytes earlier)

Commit:
Fix transition from welcome screen to AddCommic screen after rebase. Glitches fixed; add hard-coded scrolling on Dashboard xib for demo

Change color of hard coded scrolling views and direction lock

Delete GoogleAuthentication.swift

Revert back Guled AddComic changes; still not working

AddComicController working

Lucien Profile title glitch fixed

Committing changes to rebase on develop for Jenkins stuff

Commit:
UICollectionView in separate xib file added. Add separate xib for UICollectionViewCell then load xib to main UICollectionView view controller

Change size of LendingCollectionViewCell but not showing up properly. Last commit before breaking PR/branch to smaller parts

Trim extra lines

Remove GoogleAuthentication.swift

Update fonts where they were not updated previously. Change app delegate root controller back to RootViewController

Remove extra whitespace

PR changes with exception of a few

hasCollection String to Bool working

IntrepidBuildServer commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,58 @@ final class ProfileViewController: UIViewController {
+        name.font = LucienTheme.Fonts.muliRegular(size: 18.0)
+        email.font = LucienTheme.Fonts.muliRegular(size: 18.0)
+    }
+    
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Trailing Whitespace Violation: Lines should not have trailing whitespace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Atrailing_whitespace)

IntrepidBuildServer commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/In...
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/dbaf92ad77402d5c18d5db97000a3309873ed631/CameraViewController.swift#L81">CameraViewController.swift#L81</a> - <strong><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/dbaf92ad77402d5c18d5db97000a3309873ed631/lucien-ios/Camera/CameraViewController.swift#L81">lucien-ios/Camera/CameraViewController.swift#L81</a></strong>: value ‘image’ was defined but never used; consider replacing with boolean test  <br /><code>
            if let image = getImageFromSampleBuffer(buffer: sampleBuffer) {
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.370 (0.373) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 7.668 (7.670) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -23,19 +23,16 @@ final class LucienAPIClient: APIClient {
-                    return completion(.failure(LucienAPIError.cannotDecode))
-                }
-                completion(.success(user))
+                guard let authenticationResponse = try? decoder.decode(AuthenticationResponse.self, from: result) else { return }
```
Two things: one, if you return from a function with a completion without calling the completion, whatever is waiting for it will never get a response. So two, in this case instead of making it an optional `try`, instead use a regular one in a `do-catch` block and if there's an error, call the completion with it.

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -23,19 +23,16 @@ final class LucienAPIClient: APIClient {
 
     func getCurrentUser(completion: @escaping (Result<User>) -> Void) {
-        let lucienRequest = LucienRequest.getCurrentUser
+        let lucienRequest = LucienRequest.getCurrentUser()
```
No need for the parens

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+                    let user = try decoder.decode(User.self, from: result)
+                    completion(.success(user))
+                } catch(let error) {
+                    print(error)
```
Remove print

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
     }
+
+    func getDashboard(completion: @escaping (Result<Dashboard>) -> Void) {
+        let lucienRequest = LucienRequest.getDashboard()
```
No need for parens

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -13,7 +13,7 @@ enum LucienAPIError: Error {
+    func authenticateUser(code: String, completion: @escaping (Result<AuthenticationResponse>) -> Void) {
         let lucienRequest = LucienRequest.authenticate(code: code)
         let urlRequest = lucienRequest.urlRequest
         self.sendRequest(urlRequest) { response in
```
Remove `self`

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -23,19 +23,16 @@ final class LucienAPIClient: APIClient {
-        let lucienRequest = LucienRequest.getCurrentUser
+        let lucienRequest = LucienRequest.getCurrentUser()
         let urlRequest = lucienRequest.urlRequest
         self.sendRequest(urlRequest) { response in
```
Remove `self`

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -23,19 +23,16 @@ final class LucienAPIClient: APIClient {
     func getCurrentUser(completion: @escaping (Result<User>) -> Void) {
-        let lucienRequest = LucienRequest.getCurrentUser
+        let lucienRequest = LucienRequest.getCurrentUser()
         let urlRequest = lucienRequest.urlRequest
```
You can probably just say `let urlRequest = LucienRequest.getCurrentUser.urlRequest` rather than creating the `LucienRequest` separately. Same things below

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+    func getDashboard(completion: @escaping (Result<Dashboard>) -> Void) {
+        let lucienRequest = LucienRequest.getDashboard()
+        let urlRequest = lucienRequest.urlRequest
+        self.sendRequest(urlRequest) { response in
```
Remove `self`

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+        let lucienRequest = LucienRequest.getDashboard()
+        let urlRequest = lucienRequest.urlRequest
+        self.sendRequest(urlRequest) { response in
+            print(response)
```
Remove print

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+            print(response)
+            switch response {
+            case .success(let result):
+                guard let result = result else { return }
```
Must call completion on all returns

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+            case .success(let result):
+                guard let result = result else { return }
+                let decoder = JSONDecoder()
+                guard let myCollection = try? decoder.decode(Dashboard.self, from: result) else { return }
```
Must call completion on all returns

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+    func hasCollection(completion: @escaping (String) -> Void) {
+        let lucienRequest = LucienRequest.hasCollection()
+        let urlRequest = lucienRequest.urlRequest
+        self.sendRequest(urlRequest) { response in
```
Remove `self`

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+            print(response)
+            switch response {
+            case .success(let result):
+                guard let hasCollection = String(data: result!, encoding: String.Encoding.utf8) as String! else { return }
```
Must call completion on all returns

mayasaxena replied:
Don't force unwrap the result

mayasaxena replied:
Why are you casting to an implicitly unwrapped optional String here?

mayasaxena replied:
This naming is misleading. If I saw something called `hasCollection`, I'd think it was a Bool, not a string. If this endpoint returns a Bool, you should return a Bool here too

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+                guard let hasCollection = String(data: result!, encoding: String.Encoding.utf8) as String! else { return }
+                completion(hasCollection)
+            case .failure(let error):
+                print(error)
```
Must call completion here and remove print

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+        let lucienRequest = LucienRequest.hasCollection()
+        let urlRequest = lucienRequest.urlRequest
+        self.sendRequest(urlRequest) { response in
+            print(response)
```
Remove print

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -45,13 +42,48 @@ final class LucienAPIClient: APIClient {
+    }
+
+    func hasCollection(completion: @escaping (String) -> Void) {
+        let lucienRequest = LucienRequest.hasCollection()
```
No need for parens

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -10,27 +10,37 @@ import Foundation
     static let baseURL = "https://lucien-server-staging.herokuapp.com"
-    static let acceptHeader = "application/vnd.lucien-app.com; version=1"
-    static var authorizationHeader: String?
+    static let acceptHeader: String? = "application/vnd.lucien-app.com; version=1"
```
Why did you define the type explicitly here?

artsygracie17 replied:
The acceptHeader is supposed to remain the same throughout all the requests -- should I keep it this way?

mayasaxena replied:
If it stays the same, and it's not nil a) it doesn't need to be optional and b) you don't need the type declaration (`: String?`) because the compiler can infer that `acceptHeader` is a string from the definition.

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -10,27 +10,37 @@ import Foundation
-    static let acceptHeader = "application/vnd.lucien-app.com; version=1"
-    static var authorizationHeader: String?
+    static let acceptHeader: String? = "application/vnd.lucien-app.com; version=1"
+    static var authorizationHeader: String = ""
```
No need to define type explicitly, it can be inferred from the set value

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -10,27 +10,37 @@ import Foundation
 
     case authenticate(code: String)
-    case getCurrentUser
+    case getCurrentUser()
```
No need for parens here or below

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -40,13 +50,21 @@ enum LucienRequest: Request {
         switch self {
         case .authenticate, .getCurrentUser:
             return nil
+        case .getDashboard:
```
Group these with the `authenticate` and `getCurrentUser` cases

mayasaxena commented on lucien-ios/APIClient/Request.swift:
```diff
@@ -16,9 +16,10 @@ public enum HTTPMethod: String {
     static var baseURL: String { get }
-    static var acceptHeader: String { get }
-    static var authorizationHeader: String? { get }
+    static var acceptHeader: String? { get }
```
Why is this optional now?

mayasaxena commented on lucien-ios/Dashboard/LendingCollectionViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class LendingCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var cellieLabel: UILabel!
```
What's up with the naming here?

mayasaxena replied:
IBOutlets should be private

artsygracie17 replied:
Sorry about cellieLabel - it was just a cute naming convention when I was playing around how to connect the `UICollectionViewCell` to the `UICollectionView`. It should be a comic -- I'll rename it!

mayasaxena replied:
No worries! It's fine to name stuff silly names when playing around with it, just make sure it doesn't end up in the PR

artsygracie17 replied:
Since I need to change the text of the cell from the view controller, should I keep it non-private? I can't access it otherwise

mayasaxena commented on lucien-ios/Dashboard/LendingCollectionViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+class LendingCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var cellieLabel: UILabel!
+
+    override func awakeFromNib() {
```
If this only calls `super` it should be removed

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+        self.collectionView.dataSource = self
+    }
+
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
```
This returns 1 by default so it only needs to be implemented if that's different

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+import UIKit
+
+class LendingViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
```
Classes should be final by default

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    @IBOutlet weak var collectionView: UICollectionView!
+
+    let reuseIdentifier = "cell"
```
I know you're not done with implementing the feature, but this code is going to be merged into the codebase. Having fake data is okay, but the reuse identifier and other things should be correct from the start.

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.collectionView.register(UINib(nibName:"LendingCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: reuseIdentifier)
```
Remove all uses of self

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return self.objects.count
```
Remove `self`

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as? LendingCollectionViewCell
+        cell?.cellieLabel.text = self.objects[indexPath.item]
```
Custom cell UI elements shouldn't be modified in this function. Instead, create a `configure` function that takes in the values that need to be set and set them in the cell itself

mayasaxena replied:
Remove use of self

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,39 @@
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as? LendingCollectionViewCell
+        cell?.cellieLabel.text = self.objects[indexPath.item]
+
+        return cell!
```
Don't force unwrap. Dequeue the cell in a `guard` and return a `UITableViewCell` in the return.

artsygracie17 replied:
I'm having trouble converting `UICollectionViewCell` to `UITableViewCell` -- what does converting to `UITableViewCell` in the return supposed to do?

mayasaxena replied:
Whoops my bad
I mean `UICollectionViewCell`, I just saw dequeue and went to table view

mayasaxena replied:
As in, if you can't cast to the correct type of cell, just return a default one

mayasaxena commented on lucien-ios/Extensions/UINavigationBar.swift:
```diff
@@ -0,0 +1,26 @@
+
+extension UINavigationBar {
+
+    static func setNavBarBackground(navigationController: UINavigationController) {
```
Good thinking with pulling this out into an extension, but the functions here don't belong in a `UINavigationBar` extension - you're never using the actual navigation bar you're extending. Two options here: 1) make this a `UINavigationController` extension, I do NOT suggest that. 2) Get rid of the `static`, set the properties of `self` e.g.
```
isTranslucent = true
barStyle = .default
...
```
and call it like so in the class you're currently calling them in
```
navigationController?.navigationBar.setBackground()
```
Obviously it needs to be renamed a bit.

mayasaxena commented on lucien-ios/Extensions/UINavigationBar.swift:
```diff
@@ -0,0 +1,26 @@
+        navigationController.navigationBar.setValue(true, forKey: "hidesShadow")
+    }
+
+    static func setNavBarTitle(navigationController: UINavigationController, title: String) {
```
Same comments as above

mayasaxena commented on lucien-ios/Extensions/UINavigationBar.swift:
```diff
@@ -0,0 +1,26 @@
+//
+//  UINavigationBar.swift
```
File naming is off - should be `UINavigationBar+Extensions.swift`

mayasaxena commented on lucien-ios/Extensions/UIViewController.swift:
```diff
@@ -0,0 +1,26 @@
+//
+//  UIViewController.swift
```
File naming is off - should be `UIViewController+Extensions.swift`

mayasaxena commented on lucien-ios/Extensions/UIViewController.swift:
```diff
@@ -0,0 +1,26 @@
+
+extension UIViewController {
+
+//    @objc func backButtonTapped() {
```
Remove commented out code

mayasaxena commented on lucien-ios/Extensions/UIViewController.swift:
```diff
@@ -0,0 +1,26 @@
+}
+
+extension UILabel {
+    func addTextSpacing(amount: Float) {
```
I would omit the label on this parameter and rename it to `spacing`. `amount` is not descriptive enough

mayasaxena commented on lucien-ios/Extensions/UIViewController.swift:
```diff
@@ -0,0 +1,26 @@
+extension UILabel {
+    func addTextSpacing(amount: Float) {
+        if let labelText = text, labelText.characters.count > 0 {
+            let attributedString = NSMutableAttributedString(string: labelText)
```
You can simplify this to
```
let attributedString = NSAttributedString(
    string: labelText, 
    attributes: [NSAttributedStringKey.kern : spacing]
)
```

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+    let lucienAPIClient = LucienAPIClient()
+    var currentUser: User?
+    var dashboardData: Dashboard?
+    var hasCollection: String?
```
Why is this a String?

artsygracie17 replied:
Maggie was sending it as `true` or `false` from the backend -- the intention was to send me a boolean, but I tried decoding it into a Boolean and did not work. Converting it to a String works, but I can ask them to send me 0 or 1? What would you recommend?

mayasaxena replied:
Hang on I'll come over and we can debug - it should be able to decode into a Boolean

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private func configureNavigationController() {
+        UINavigationBar.setNavBarBackground(navigationController: navigationController!)
```
See my comments on the `UINavigationBar` extension, and don't force unwrap

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    private func configureNavigationController() {
+        UINavigationBar.setNavBarBackground(navigationController: navigationController!)
+        self.navigationItem.title = "Lucien"
```
Remove self

mayasaxena commented on lucien-ios/LoginScreen/RootViewController.swift:
```diff
@@ -50,10 +56,34 @@ final class RootViewController: UIViewController, GIDSignInDelegate, GIDSignInUI
         guard let idToken = user.authentication.idToken else { return }
-        loginViewModel.authenticateUser(code: idToken)
+        loginViewModel.authenticateUser(code: idToken) {
+            //            self.loginViewModel.getDashboard {
```
Remove commented out code

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,58 @@ final class ProfileViewController: UIViewController {
+    }
+
+    private func configureNavigationController() {
+        guard self.navigationController != nil else { return }
```
Same comments as the other configure function

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,58 @@ final class ProfileViewController: UIViewController {
+    }
+
+    @objc func back(sender: UIBarButtonItem) {
+        self.navigationController?.popViewController(animated: true)
```
Remove self

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,58 @@ final class ProfileViewController: UIViewController {
+    private func setNavBarBackButton() {
+        let backButton = UIBarButtonItem(image: UIImage(named: "navBackButton"), style: .plain, target: self, action: #selector(back(sender:)))
+        backButton.tintColor = LucienTheme.dark
+        self.navigationItem.leftBarButtonItem = backButton
```
Remove self

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
 
 final class StartMyCollectionViewController: UIViewController {
+
+    @IBOutlet weak var addBook: UIButton!
```
This should be named `addBookButton`

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
 final class StartMyCollectionViewController: UIViewController {
+
+    @IBOutlet weak var addBook: UIButton!
+    @IBOutlet weak var welcomeImage: UIImageView!
```
This should be named `welcomeImageView`

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
+
+    @IBOutlet weak var addBook: UIButton!
+    @IBOutlet weak var welcomeImage: UIImageView!
+    @IBOutlet weak var toDashboard: UIButton!
```
This should be named `toDashboardButton`

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
+    @IBOutlet weak var addBook: UIButton!
+    @IBOutlet weak var welcomeImage: UIImageView!
+    @IBOutlet weak var toDashboard: UIButton!
+    @IBOutlet weak var getStartedDescription: UILabel!
```
Same as above, add label to end

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
+    @IBOutlet weak var welcomeImage: UIImageView!
+    @IBOutlet weak var toDashboard: UIButton!
+    @IBOutlet weak var getStartedDescription: UILabel!
+    @IBOutlet weak var welcome: UILabel!
```
Same as above

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.432 (0.434) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 11.208 (11.211) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -13,42 +13,83 @@ enum LucienAPIError: Error {
+                    return completion(.failure(LucienAPIError.noResult))
+                }
+                 let resultBool = (resultString as NSString).boolValue
+                print("hasCollection result: \(resultBool)")
```
Remove print

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -13,42 +13,83 @@ enum LucienAPIError: Error {
+                guard let resultString = String(data: result, encoding: String.Encoding.utf8) else {
+                    return completion(.failure(LucienAPIError.noResult))
+                }
+                 let resultBool = (resultString as NSString).boolValue
```
There's a better name for this that's more descriptive - `hasCollection`

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -11,26 +11,36 @@ enum LucienRequest: Request {
     static let baseURL = "https://lucien-server-staging.herokuapp.com"
     static let acceptHeader = "application/vnd.lucien-app.com; version=1"
-    static var authorizationHeader: String?
+    static var authorizationHeader: String = ""
```
The compiler can infer the `: String` part

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+final class DashboardViewController: UIViewController {
+
+    @IBOutlet weak var dashboardScrollView: UIScrollView!
```
Should be private

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        print("In Dashboard VC")
```
Remove print

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    private func configureNavigationController() {
+        navigationController?.navigationBar.setNavBarBackground()
+        self.navigationItem.title = "Lucien"
```
remove `self`

mayasaxena commented on lucien-ios/Dashboard/DashboardViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+    @objc private func viewProfileButtonPressed() {
+        let viewProfileController = ProfileViewController()
+        self.navigationController?.pushViewController(viewProfileController, animated: true)
```
Remove `self`

mayasaxena commented on lucien-ios/Dashboard/LendingCollectionViewCell.swift:
```diff
@@ -0,0 +1,13 @@
+
+import UIKit
+
+class LendingCollectionViewCell: UICollectionViewCell {
```
should be final

mayasaxena commented on lucien-ios/Dashboard/LendingCollectionViewCell.swift:
```diff
@@ -0,0 +1,13 @@
+import UIKit
+
+class LendingCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var lendedComicTitle: UILabel!
```
should be private

mayasaxena replied:
Should be suffixed with `label` e.g. `lendedComicTitleLabel`. Can also probably shorten to `titleLabel` since there aren't any other titles in the cell

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+
+final class LendingViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
+
+    @IBOutlet weak var collectionView: UICollectionView!
```
should be private

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+
+    @IBOutlet weak var collectionView: UICollectionView!
+
+    let reuseIdentifier = "comicCell"
```
This should be defined within the cell rather than here

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        collectionView.register(UINib(nibName:"LendingCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: reuseIdentifier)
+        collectionView.delegate = self
```
These can be set in the nib

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        collectionView.register(UINib(nibName:"LendingCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: reuseIdentifier)
```
String shouldn't be hardcoded. Instead, use `String(describing: LendingCollectionViewCell.self)`

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath)
+        guard let lendingCell = cell as? LendingCollectionViewCell else { return cell}
+//        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as! LendingCollectionViewCell else { return }
```
remove commented code

mayasaxena commented on lucien-ios/Dashboard/LendingViewController.swift:
```diff
@@ -0,0 +1,40 @@
+        return lendingCell
+    }
+
+    func configure(cell: LendingCollectionViewCell, lendedComicTitle: String) {
```
This should be a function within `LendingCollectionViewCell` so that `lendedComicTitle` can be private

mayasaxena commented on lucien-ios/Extensions/UIViewController+Extensions.swift:
```diff
@@ -0,0 +1,25 @@
+
+import Foundation
+
+extension UIViewController {
```
Remove empty extension

mayasaxena commented on lucien-ios/Extensions/UIViewController+Extensions.swift:
```diff
@@ -0,0 +1,25 @@
+
+}
+
+extension UILabel {
```
Since this is a `UILabel` extension, it should live in a file called `UILabel+Extensions`, not here

mayasaxena commented on lucien-ios/LoginScreen/RootViewController.swift:
```diff
@@ -50,10 +56,29 @@ final class RootViewController: UIViewController, GIDSignInDelegate, GIDSignInUI
-        loginViewModel.authenticateUser(code: idToken)
+        loginViewModel.authenticateUser(code: idToken) {
+            self.loginViewModel.hasCollection {
+                print(self.loginViewModel.hasCollection)
```
Remove print

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,56 @@ final class ProfileViewController: UIViewController {
     @IBOutlet private weak var name: UILabel!
```
These labels should have `label` appended to their names, otherwise it looks like they're data and not UI elements.

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,56 @@ final class ProfileViewController: UIViewController {
+        configureNavigationController()
+        loginViewModel.getCurrentUser {
+            guard let currentUser = self.loginViewModel.currentUser else { return }
+            debugPrint("in debug print \(currentUser)")
```
Remove print

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -16,18 +16,56 @@ final class ProfileViewController: UIViewController {
+        loginViewModel.getCurrentUser {
+            guard let currentUser = self.loginViewModel.currentUser else { return }
+            debugPrint("in debug print \(currentUser)")
+            self.name.text = currentUser.firstName + " " + currentUser.lastName
```
The name and email should be properties in the view model that are set when the user data is fetched. Thus, `getCurrentUser` doesn't need a completion

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,63 @@
+    var dashboardData: Dashboard?
+    var hasCollection: Bool = false
+
+    func getCurrentUser(completion: @escaping () -> Void) {
```
Why does this need a completion?

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    func authenticateUser(code: String, completion: @escaping () -> Void) {
```
This is okay to have a completion since we're not using Rx yet but once we do it shouldn't have one

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    func getDashboard(completion: @escaping () -> Void) {
```
Why does this have a completion? Also if you're going to have a completion, you need to call it from all exit paths (it's not being called in the failure case). This applies to all completions, in this file and in general

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    func hasCollection(completion: @escaping () -> Void) {
```
This completion isn't doing anything. The point of a completion is to act as a callback when an action completes, and to return any information needed from the action. In this case, the info we need from the action is `hasCollection` so the completion should be of type `(Bool) -> Void` and it should return the result on success. If you don't know why, I can clarify

mayasaxena commented on lucien-ios/LoginScreen/RootViewController.swift:
```diff
@@ -50,10 +56,29 @@ final class RootViewController: UIViewController, GIDSignInDelegate, GIDSignInUI
+        loginViewModel.authenticateUser(code: idToken) {
+            self.loginViewModel.hasCollection {
+                print(self.loginViewModel.hasCollection)
+                self.loginViewModel.hasCollection ? self.showDashboard() : self.showStartMyCollection()
```
This statement isn't quite right - we're in a callback from a function that should retrieve a Bool from the server. We shouldn't need to go into the view model to get that Bool, it should be part of the completion. Eventually we'll get rid of the completions entirely with Rx

mayasaxena commented on lucien-ios/Models/User.swift:
```diff
@@ -31,15 +32,18 @@ struct User: Decodable {
+        do {
+            googlePictureURL = try userContainer.decode(String.self, forKey: .googlePictureURL)
+        } catch _ {
+            googlePictureURL = "eeerrs"
```
??

mayasaxena commented on lucien-ios/Models/User.swift:
```diff
@@ -31,15 +32,18 @@ struct User: Decodable {
         lastName = try userContainer.decode(String.self, forKey: .lastName)
         email = try userContainer.decode(String.self, forKey: .email)
-        googlePictureURL = try userContainer.decode(String.self, forKey: .googlePictureURL)
+        do {
```
Why is this in a `do-catch`?

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
+    }
+
+    func setUpStyling() {
+        welcomeLabel.font = LucienTheme.Fonts.muliExtraLightItalic(size: CGFloat(48.0))
```
Each chunk of lines starting with the same label should be grouped into a smaller function called `setup<whatever it's setting up>` e.g. `setupWelcomeLabel`

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,48 @@
+    @IBOutlet private weak var addBookButton: UIButton!
+    @IBOutlet private weak var welcomeImageView: UIImageView!
+    @IBOutlet private weak var toDashboardButton: UIButton!
+    @IBOutlet private weak var getStartedDescription: UILabel!
```
This needs to have `Label` appended to the end, otherwise it looks like a String or other data type

Commit:
PR changes and create DashboardViewModel to put getDashboard and other functionality in there

Remove do-catch in User model and replace with String optional instead

PR changes round 2

PR changes round 2

Add completions to LoginViewModel

IntrepidBuildServer commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+        welcomeLabel.textColor = LucienTheme.dark
+        welcomeLabel.addTextSpacing(spacing: 1.2)
+    }
+    
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Trailing Whitespace Violation: Lines should not have trailing whitespace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Atrailing_whitespace)

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_WPXBFUY72OYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.400 (0.403) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 51.887 (51.891) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_WPXBFUY72OYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.427 (0.429) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.545 (6.548) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/Dashboard/DashboardViewModel.swift:
```diff
@@ -0,0 +1,28 @@
+                self.dashboardData = result
+                completion()
+            case .failure(let error):
+                print(error)
```
Call `completion` here

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+                LucienRequest.authorizationHeader = "Bearer " + authenticationResponse.token
+                completion()
+            case .failure(let error):
+                print(error)
```
Call completion here

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+            case .success(let result):
+                completion(result)
+            case .failure(let error):
+                print(error)
```
Call completion here

mayasaxena commented on lucien-ios/Models/Dashboard.swift:
```diff
@@ -0,0 +1,27 @@
+        case myCollectionCount = "my_collection_count_by_issue"
+    }
+
+    init(from decoder: Decoder) throws {
```
This is fine for now, but we can potentially make this a `DashboardResponse` class that decodes a `Dashboard` in the same way we do with `User`. As long as `Dashboard` is defined outside of `DashboardResponse` we'll be okay

artsygracie17 replied:
In my `gracie/dashboard-lending` that's the way I have it!

mayasaxena replied:
Perfect, you can leave it for this PR then

mayasaxena commented on lucien-ios/ProfileViewController.swift:
```diff
@@ -12,22 +12,59 @@ final class ProfileViewController: UIViewController {
-        name.textColor = LucienTheme.dark
-        email.textColor = LucienTheme.dark
-        profileTitle.textColor = LucienTheme.dark
+        nameTitleLabel.textColor = LucienTheme.coolGrey
```
You should be able to set all of this in Interface Builder instead of programmatically. If you need help with that let me know

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+    }
+
+    func setUpWelcomeLabel() {
+        welcomeLabel.font = LucienTheme.Fonts.muliExtraLightItalic(size: CGFloat(48.0))
```
All of this but the text spacing can be set in IB

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+    }
+    
+    func setUpGetStartedDescriptionLabel() {
+        getStartedDescriptionLabel.textColor = LucienTheme.dark
```
All of this can be set in IB

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+    }
+
+    func setUpToDashboardButton() {
+        toDashboardButton.titleLabel?.addTextSpacing(spacing: 1.0)
```
All of this can be set in IB

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+
+    func setUpToDashboardButton() {
+        toDashboardButton.titleLabel?.addTextSpacing(spacing: 1.0)
+        toDashboardButton.setTitleColor(LucienTheme.dark, for: .normal)
```
All of this but the `layer` stuff can be set in IB

mayasaxena commented on lucien-ios/StartMyCollection/StartMyCollectionViewController.swift:
```diff
@@ -9,9 +9,64 @@
+    }
+
+    func setUpAddBookButton() {
+        addBookButton.titleLabel?.addTextSpacing(spacing: 1.0)
```
Same as above, set in IB

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.308 (0.310) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.048 (6.051) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_WPXBFUY72OYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.374 (0.376) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.223 (6.225) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.363 (0.365) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 10.328 (10.330) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_WPXBFUY72OYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.369 (0.371) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 7.349 (7.351) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Commit:
Set navigation title (somehow lost from previous rebase) in profile view controller

ghost commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-12-LNUW4FSCNJOPVFDXBXBMVQPE64BNH2YTZNJIGTXBYCKFUJ3SYO7Q/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_WPXBFUY72OYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.368 (0.373) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 8.019 (8.021) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
